spec/consensus/proposer-based-timestamp/pbts-algorithm_001_draft.md

```diff
@@ -0,0 +1,136 @@
+# Proposer-Based Time - Part II
+
+## Updated Consensus Algorithm
+
+### Outline
+
+The algorithm in the [arXiv paper][arXiv] evaluates rules of the received messages without making explicit how these messages are received. In our solution, we will make some message filtering explicit. We will assume that there are message reception steps (where messages are received and possibly stored locally for later evaluation of rules) and processing steps (the latter roughly as described in a way similar to the pseudo code of the arXiv paper).
+
+In contrast to the original algorithm the field `proposal` in the `PROPOSE` message is a pair `(v, time)`, of the proposed consensus value `v` and the proposed time `time`. 
+
+#### **[PBTS-RECEPTION-STEP.0]**
+
+In the reception step at process `p` at local time `now_p`, upon receiving a message `m`:
+- if the message `m` is of type `PROPOSE` and satisfies `now_p - PRECISION <  m.time < now_p + PRECISION + MSGDELAY`, then mark the message as `timely`
+> if `m` does not satisfy the constraint consider it `untimely`
+ 
+
+#### **[PBTS-PROCESSING-STEP.0]**
+In the processing step, based on the messages stored, the rules of the algorithms are
+executed. Note that the processing step only operates on messages
+for the current height. The consensus algorithm rules are defined by the following updates to arXiv paper.
+
+#### New `StartRound`
+
+There are two additions
+- in case the proposer's local time is smaller than the time of the previous block, the proposer waits until this is not the case anymore (to ensure the block time is monotonically increasing)
+- the proposer sends its time `now_p` as part of its proposal
+
+#### **[PBTS-ALG-STARTROUND.0]**
+```go
+function StartRound(round) {
+  round_p ← round
+  step_p ← propose
+  if proposer(h_p, round_p) = p {
+    wait until now_p > block time of block h_p - 1 // new wait condition
+    if validValue_p != nil {
+      proposal ← (validValue_p, now_p) // added "now_p"
+    }
+    else {
+      proposal ← (getValue(), now_p)   // added "now_p"
+    }
+    broadcast ⟨PROPOSAL, h_p, round_p, proposal, validRound_p⟩
+  }
+  else {
+    schedule OnTimeoutPropose(h_p,round_p) to be executed after timeoutPropose(round_p)
+  }
+}
+```
+
+#### New Rule Replacing Lines 22 - 27
+
+- a validator prevotes for the consensus value `v` **and** the time `t`
+- the code changes as the `PROPOSAL` message carries time (while `lockedValue` does not)
+
+#### **[PBTS-ALG-UPON-PROP.0]**
+```go
+upon timely(⟨PROPOSAL, h_p, round_p, (v,t), −1⟩) from proposer(h_p, round_p) while step_p = propose do {
```

Comment:
What is this timely function?

Reply:
In the [reception step](https://github.com/tendermint/spec/blob/ilina/proposer-based-timestamp/spec/consensus/proposer-based-timestamp/pbts_001_draft.md#reception-step) we mark a `propose` message as timely if it arrives in a given interval. This just expresses that this was the case for that message. 

Does that help?  Do you think we should write that in a more clear way?

Reply:
That helps! I think I just missed the definition in reading the raw markdown, its pretty clear in the rendered version =)